openssl/sha.go

```diff
@@ -9,48 +9,122 @@ package openssl
 // #include "goopenssl.h"
 import "C"
 import (
+	"crypto"
 	"errors"
 	"hash"
+	"runtime"
+	"strconv"
 	"unsafe"
 )
 
-// NewSHA1 returns a new SHA1 hash.
-func NewSHA1() hash.Hash {
-	h := new(sha1Hash)
+type evpHash struct {
+	md        *C.EVP_MD
+	ctx       *C.EVP_MD_CTX
+	ctx2      *C.EVP_MD_CTX
+	size      int
+	blockSize int
```

Comment:
```c
#define OSSL_CIPHER_PARAM_BLOCK_SIZE           "blocksize"    /* size_t */
```

This seems to indicate that `blockSize` should be a native int vs. a 32 bit int. I was looking around for the best way to marshal a `size_t` in Golang but didn't see any great examples. 

I'm also not very familiar with Go C interop rules so is there something I'm missing here?

Reply:
Good catch. In Go, `int` is a signed native integer, not a 32bit integer. But the mismatch is still there, as `size_t` is an unsigned native integer. Go uses `int` instead of `uint`  to represent length and capacity. see [the spec](https://go.dev/ref/spec#Length_and_capacity):

> The built-in functions len and cap take arguments of various types and return a result of type int. The implementation guarantees that the result always fits into an int.

So it is a common idiom to just cast `size_t` to `int`, and vise versa, assuming that casting from `size_t` to `int` can overflow, whereas casting from `int` to `size_t` is safe.

Having said this, in this particular case we are hardcoding `blockSize` instead of taking it from OpenSSL, as this value is static and known when creating a sha object, i.e. `sha1` block size is always 64.

On the other hand, there a a bunch of places where we are casting `size_t` to `int`. I'll check if they can overflow and fix it if necessary in this same PR.

Reply:
> Go, int is a signed native integer, not a 32bit integer. 

Interesting. Did not realize that.